### ISSUE TITLE
Clasp can't seem to see bound script projects

### ISSUE THREAD

**Issue by ShowMeBillyJo** [google/clasp#208]
## Expected Behavior
`clasp list` should show all of my script projects, including those bound to G Drive documents like spreadsheets.

## Actual Behavior
`clasp list` doesn't show bound script projects.

The `FileSubmission` project is owned by someone else and shared with me.

![image](https://user-images.githubusercontent.com/179590/40895353-8ab9b770-6774-11e8-887b-d5f448498452.png)

## Steps to Reproduce the Problem

  1. Create a spreadsheet document in the G Drive of the same account you're logged into Clasp with.
  1. Add a script project to the spreadsheet.
  1. `clasp list`

## Specifications

  - Node version 10.3.0
  - Clasp version 1.3.1
  - Windows 10 x64 v 1803

**Comment by grant** [comment]
I think we're currently limited by the API on this one.
- `clasp list` uses the Drive API
  - which maybe can't detect bounded scripts?
- The Apps Script API doesn't have a method for listing projects.

Let me know if you see the scripts with the "Try this API" function on this page:
https://developers.google.com/drive/api/v3/reference/files/list

Query:
https://github.com/google/clasp/blob/master/src/commands.ts#L157

**Comment by bhaelochon** [comment]
No dice. :cry: All that gets returned is its container:

```
{
  "kind": "drive#file",
  "id": "<snip>",
  "name": "XP Logbook",
  "mimeType": "application/vnd.google-apps.spreadsheet"
}
```

Just to be sure, I grepped the output for `application/vnd.google-apps.script`, but nothing.

**Comment by virusakos** [comment]
I am also interested in container (bound) scripts and I was looking for a way to automate it with clasp.

https://developers.google.com/apps-script/guides/bound

"Creating a bound script

To create a bound script, open a Google Sheets, Docs, or Forms file, then select Tools > Script editor. To reopen the script in the future, do the same thing. Because bound scripts do not appear in Google Drive, that menu is the only way to find or open the script."

I have looked into Google Sheet API, Google Drive API and Google App Scripts API for a way to get the container (bound) script id but I have not found anything.

**Comment by bhaelochon** [comment]
That documentation isn't entirely correct. You can also get to bound scripts at https://script.google.com.

Cloning and pulling bound scripts still works, but the URL method of getting the script ID is unreliable. If you open your script project from https://script.google.com, it's fine. If you open it from `Tools -> Script editor` in the bound document, the URL doesn't match the expected format.

Expected: `https://script.google.com/d/<SCRIPT_ID>/edit`

Actual: `https://script.google.com/macros/d/<MYSTERY_ID>/edit`

The only sure way of getting the right value is from `File -> Project properties -> Info -> Script ID`.

**Comment by grant** [comment]
@ShowMeBillyJo Can you file an issue against the Drive API so the team can eventually implement this?

https://issuetracker.google.com/issues/new?component=191650&template=824106

Then please update the above issue comment with a link to the bug.

**Comment by grant** [comment]
If anyone could create an issue with the above link, and then comment, that would be very helpful to make progress on this bug.

**Comment by bobbytait** [comment]
Work-around for existing projects:

As @ShowMeBillyJo mentioned, you can get the correct script ID from https://script.google.com. Then I used this ID to replace the one in my .clasp.json file, and it seems to be working.

**Comment by grant** [comment]
To be clear, you can clone bound scripts, but just not list them.

Example:

`clasp clone "https://script.google.com/d/fWc1OG320D82xo5wYVfqivevzZEItbPP3v7l1eh7XQTEm3vZe3YZpjwuD/edit"`

Source for Drive List:

https://github.com/google/clasp/blob/master/src/commands/list.ts#L25

**Comment by jobdiogenes** [comment]
Have someone Issue a feature request? I did not found any.
If so, them link here.

**Comment by grant** [comment]
> Have someone Issue a feature request? I did not found any.
> If so, them link here.

Nobody has filed a FR yet.

**Comment by jobdiogenes** [comment]
Now issued an Feature Request to Drive Api.
(https://issuetracker.google.com/issues/139438152)

I ask to include scripts projetcs inside Docs, etc.  in drive.file.list

I also suggest to make this files external and make then only linked to Script projects.
In that way, one Project could to linked to many Forms, Docs, etc. and may be one file be linked to many script projects. Leaving the need to make Addons.

**Comment by symbl-ay** [comment]
The ticket has been marked as will not fix. It might be time to close this.

Creating a project `--type=api` and then using it from linked Drive documents appears to be the use case that was intended.

**Comment by ShowMeBillyJo** [comment]
Google ticket to support this closed as wontfix

**Comment by MatthewGitter** [comment]
As of 11/21/24 clasp can list bound projects along with standalone projects, and it can clone and pull bound projects as it can standalone projects. However, the proper permissions scopes must be selected during clasp login in the authentication flow in the browser on the page introducing the list with "Select what clasp – The Apps Script CLI can access":

clasp list requires "See information about your Google Drive files. [✓]".
clasp clone <script ID> (where you copy <script ID> from the GAS project's Project Settings page) as well as clasp pull and clasp push require "Create and update Google Apps Script projects. [✓]" (also ensure that the local dir does not contain a .clasp.json file already, as the clone operation needs to overwrite it but will be blocked by an existing one).
clasp push additionally requires activating Google Apps Script API at https://script.google.com/u/5/home/usersettings , which can take a few minutes to take effect for clasp to use it.

**Comment by albertored** [comment]
> As of 11/21/24 clasp can list bound projects along with standalone projects, and it can clone and pull bound projects as it can standalone projects

Are you sure? For me it is still failing

**Comment by MatthewGitter** [comment]
> > As of 11/21/24 clasp can list bound projects along with standalone projects, and it can clone and pull bound projects as it can standalone projects
> 
> [@MatthewGitter](https://github.com/MatthewGitter) are you sure? For me it is still failing

You have set the configs I specified in the comment you quoted?

**Comment by albertored** [comment]
> 1. You have set the configs I specified in the comment you quoted?

Yes, on clasp login I checked all the available permission present in the opened web page. Among them there are also the one you are referring.

> 2. If you have, what is clasp doing and outputting instead of the expected successful result?

```
$ clasp list
No script files found.

$ clasp clone <script_id>
Could not find script.
Did you provide the correct scriptId?
Are you logged in to the correct account with the script?
```

**Comment by MatthewGitter** [comment]
I just tested it OK. Did you follow the procedure in the Script Guide, as I reiterate in this message?

* [Home > Google Workspace > Apps > Script Guides > Use the command line interface with clasp](https://developers.google.com/apps-script/guides/clasp):

Please let me know any deviation from what you did or what you got as results.

Here is how that procedure results in the failure you reported:
`$ clasp login`

If you are already logged into your Google account that you're logging into with clasp, you get:
```
Warning: You seem to already be logged in *globally*. You have a ~/.clasprc.json
Logging in globally…
```

 Regardless of that, you will get:
```
 🔑 Authorize clasp by visiting this url:
 https://accounts.google.com/o/oauth2/v2/auth[...]
```
That (full) URL will open in your Chrome browser, prompting you to login and authorize scopes for the clasp you invoked to use (and to retain until it expires). You should select only the scopes necessary to authorize for the development you expect.

**THE ERROR YOU REPORTED RESULTS IF YOU DO NOT SELECT:**
`Create and update Google Apps Script projects. ☑`

When the page reports your login was successful, you close that browser tab. Then you will see in response to the `clasp login` (<username> populated with your local username):
```
Authorization successful.

Default credentials saved to: /home/<username>/.clasprc.json.
```

Then in the bound GAS script's Editor's Settings tab's IDs / Script ID section copy the Script ID. Next run (with your script ID of course):
`$ clasp clone <scriptID>`

If you didn't authorize `Create and update Google Apps Script projects. ☑` you get:
```
⠋ Cloning files…(node:[...]
Could not find script.
Did you provide the correct scriptId?
Are you logged in to the correct account with the script?
```

If you did authorize that scope, then you will get:
```
⠋ Cloning files…(node:[...]
Cloned <N> files.
└─ <local path to cloned file 1>
[...]
└─ <local path to cloned file N>
Not ignored files: [reiterating the cloned ones if each of them was cloned OK]
└─ <local path to not ignored file 1>
[...]
└─ <local path to not ignored file N>

Ignored files: [if any]
└─ <local path to ignored file 1>
[...]
└─ <local path to ignored file N>
```

Please reply with how that procedure compares with what you're doing, or if following it resulted in success.
